src/util/converter/factory.js

```diff
@@ -0,0 +1,146 @@
+ * @param {string} text - The text to convert.
+ * @param {object} [config] - An optional config object that may contain processor specific configuration.
+ * @returns {string} - Returns the converted text.
+ */
+
+/**
+ * Defines a converter processor.
+ * @typedef {object} converterProcessor
+ * @property {string} name - The name of the converter processor. It needs to be unique within the collection.
+ * @property {converterProcessorRuntime} convert - The processor runtime that will actually convert the text.
+ */
+
+/**
+ * Checks a converter processor is valid.
+ * @param {converterProcessor} processor - The converter processor to validate.
+ * @returns {boolean} - Returns `true` if the converter processor is valid ; returns `false` otherwise.
+ */
+function validateProcessor(processor) {
+    if (!_.isPlainObject(processor)) {
+        return false;
+    }
+
+    if ('string' !== typeof processor.name || !processor.name) {
+        return false;
+    }
+
+    return 'function' === typeof processor.convert;
+}
+
+/**
+ * Creates a text converter.
+ * @param {converterProcessor[]} builtinProcessors - A list of built-in converter processors.
+ * @param {object} [builtinConfig] - An optional default config object that may contain processor specific configuration.
+ *                                   It will be forwarded to each call to the converter.
+ * @returns {converter} - Returns the text converter, ready for use.
+ * @export 'util/converter/factory'
+ */
+export default function converterFactory(builtinProcessors = [], builtinConfig = {}) {
+    let processors = [];
+
+    // find the index of the named processor
+    const findProcessor = name => processors.findIndex(processor => processor.name === name);
+
+    /**
+     * @typedef {object} converter
+     */
+    const converter = {
+        /**
+         * Converts a text with respect to the registered converter processors.
+         * @param {string} text - The text to convert.
+         * @param {object} [config] - An optional config object that may contain processor specific configuration.
+         *                            It will be merged with the possible builtin config.
+         * @returns {string} - Returns the converted text.
+         */
+        convert(text, config = {}) {
+            const localConfig = Object.assign({}, builtinConfig, config);
+            for (const processor of processors) {
+                text = processor.convert.call(converter, text, localConfig);
+            }
+
+            return text;
+        },
+
+        /**
+         * Registers a converter processor.
+         * @param {converterProcessor} processor - The converter processor to register.
+         * @returns {converter} - Chains the instance.
+         */
+        register(processor) {
+            if (!validateProcessor(processor)) {
+                throw new TypeError('The given processor is not valid!');
+            }
+
+            if (findProcessor(processor.name) > -1) {
```

Comment:
Can we get rid of `findProcessor` helper function by just moving its code to `isRegistered` method and then calling it from here?

Reply:
I was hesitating to do so indeed. I built the API iteratively, and the internal helper was used differently as I added the `isRegistered` late for easing the testing. I'll do this small refactoring.